### PR TITLE
Add new file option php_namespace.

### PR DESCRIPTION
**Pull request by TeBoring**
Use this option to change the namespace of php generated classes.
Default is empty. When this option is empty, the package name will be
used for determining the namespace.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by TeBoring** [comment]
#3122

**Comment by TeBoring** [comment]
@michaelbausor

File: php/tests/array_test.php [hunk 1]
```diff
@@ -798,153 +798,153 @@ public function testMessageSetStringFail()
...
-        $start = memory_get_usage();
-        unset($arr);
-
-        // Explicitly trigger garbage collection.
-        gc_collect_cycles();
-
-        $end = memory_get_usage();
-        $this->assertLessThan($start, $end);
-    }
+#     /**
```
Comment (michaelbausor) [inline php/tests/array_test.php]: Are these tests supposed to be commented out?

**Comment by TeBoring** [comment]
Thanks for the catch.

File: php/tests/array_test.php (hunk 1 above)
Comment (TeBoring) [inline php/tests/array_test.php]: Fixed.

**Comment by michaelbausor** [comment]
Testing locally, I am finding that this does not change the namespace of the GPBMetadata files. Is that intentional? I do not know of strong arguments either for changing it or not.

**Comment by TeBoring** [comment]
I don't think we need to change the namespace of GPBMetadata files. These files are transparent to users.

**Comment by TeBoring** [comment]
BTW, the namespace of GPBMetadata files are decided by the path of proto file instead of the package name.

**Comment by heyanlong** [comment]
@TeBoring What time php_namespace. protoc 3.4?

**Pull request merged**